### ISSUE TITLE
List available steps in project

### ISSUE THREAD

**Issue by CDFriend** [getgauge/gauge#1853]
**Is your feature request related to a problem? Please describe.**

When writing test specifications, it can be difficult to know if I have already implemented a particular step, especially when my step implementations are divided up into several files. 

**Describe the solution you'd like**

It would be very helpful to have a complete list of the steps I have implemented in a project so I don't end up re-implementing a step I already have. This list could be accessed from IDE plugins, and through the Gauge CLI. 

For example, `gauge list --steps specs/` would find all the steps referenced in the specs folder, confirm they are implemented, and return a list to the user. 

**Describe alternatives you've considered**

Currently I can use autocompletion in the VS code plugin to check if a similar step has been implemented, but this requires me to guess the first few characters of the step description. 

![2021-01-14 11_10_51-PublicUser spec - RCWebviewGauge - Visual Studio Code](https://user-images.githubusercontent.com/8661121/104637336-84802200-5659-11eb-9aec-386e8c81427d.png)

**Comment by sriv** [comment]
Thanks for the idea @CDFriend. The reason we did not include this in the `gauge list` command was because the number of steps can be quite high. 

A while ago I was wondering if there can be a mechanism to detect duplicate steps (either using the step text or the implementation). Although, I haven't investigated it much.

**Comment by CDFriend** [comment]
Understandable! That'd be a lot of information to come out of one command. It could definitely still be useful if passed into grep or a similar tool though - that way a user could quickly search through steps on the command line. 

Ultimately what I'd love to see is a feature in the VS Code plugin where we can see all steps in a list in the UI and filter through them by name. Initially I thought the "get all steps" functionality wasn't available in Gauge core but it looks like we can access that information through the gRPC API. Maybe this is a feature that should be requested in the gauge-vscode repository.

```
  /// [GetAllStepsRequest](#gauge.messages.GetAllStepsRequest)
  GetAllStepsRequest allStepsRequest = 7;
  /// [GetAllStepsResponse](#gauge.messages.GetAllStepsResponse)
  GetAllStepsResponse allStepsResponse = 8;
```

**Comment by sriv** [comment]
You're right. `GetAllStepsRequest` is available via the gauge API. However gauge-vscode does not use the grpc API, it uses the Language Server Protocol which in turn uses json-rpc under the hood.

That said, there's no reason why this cannot be added. The challenge is to get a good UX. Where would you like to see this information?

**Comment by CDFriend** [comment]
If you'll excuse my MS paint mockup...

![gauge-steps-mockup](https://user-images.githubusercontent.com/8661121/105097517-91d44c80-5a5d-11eb-8282-39b7719c8de6.png)

In essence, we were thinking about having a separate view similar to the Gauge test explorer where you could search through steps. Clicking on any given step would take you to the implementation. 

My colleague also suggested grouping steps together in a tree by their alias. For example:
```
Given I am logged in
> And I am logged in
> Log in
```

**Comment by sriv** [comment]
Let me see if there's a way to wire up such a view with vscode. Will then need to figure out who is responsible for the search/filter (vscode or gauge language server).

On a different note, would you be interested in sending a contribution (I am happy to help in any way I can)?

**Comment by CDFriend** [comment]
Absolutely - I'm happy to contribute some code for this feature. Just hoping to establish a design before getting too far into building things. 

I can start digging into the search/filter part of this in the near future.

**Comment by sriv** [comment]
What are.your thoughts on VSCode's Symbol Search - https://code.visualstudio.com/docs/editor/editingevolved#_open-symbol-by-name

If we follow VSCode's UX, Steps could be symbols that can be searched and navigated to via Symbol Search.

Implementationwise this would mean hooking up a [`WorkspaceSymbolProvider`](https://code.visualstudio.com/api/references/vscode-api#WorkspaceSymbolProvider)

**Comment by CDFriend** [comment]
Sounds like a great idea. From what I can see in VS code, we can currently search through Gauge scenarios in VS code, but not Gauge steps.

**Comment by sriv** [comment]
I had some time, so tried to look into this a bit further. 

So, we have 

- a way to retrieve all the step texts (without the location) from the implementations, 
- a way to retrieve location of a step, given it's text
- we do not have a way to retrieve _all implemented steps with their location_

The location becomes relevant because in the symbol search box, the action is to navigate to the symbol (i.e. search + select + Enter).

My first gut was to pick the steps from the spec files, but that would be incorrect for two reasons:
- a step can be used multiple times, hence there'll be a lot of duplicates
- it will not include unused steps
